### PR TITLE
Use json encoder/decoder. This should be bits faster.

### PR DESCRIPTION
**Pull request by mattn**
Marshal/Unmarshal allocate bytes to parse. but stream is not.

### CODE REVIEW

**Comment by cubicdaiya** [comment]
@mattn Thank you for the patch.

But it looks wrong. Because `sendResponse()` always returns HTTP OK(200).


For example, [POST /push](https://github.com/mercari/gaurun/blob/master/SPEC.md#post-push) allows only a POST method. If a client sends a HTTP GET request, `gaurun` returns HTTP Bad Request(400). But `gaurun` always returns HTTP OK(200) after your change is applied.

```console
classic-club@bokko% curl -I http://127.0.0.1:1056/push
HTTP/1.1 200 OK
Content-Type: application/json
Server: Gaurun 0.3.0
Date: Mon, 22 Jun 2015 20:42:13 GMT
Content-Length: 34

classic-club@bokko% curl http://127.0.0.1:1056/push   
{"message":"method must be POST"}
classic-club@bokko%
```

**Comment by cubicdaiya** [comment]
Switching `Unmarshal` to `NewDecoder` is already LGTM.

**Comment by mattn** [comment]
fixed

**Comment by cubicdaiya** [comment]
Hi @mattn , 

```go
	w.WriteHeader(code)
 	respGaurun.Message = msg
	err := json.NewEncoder(w).Encode(&respGaurun)
 	if err != nil {
 		msg := "Response-body could not be created"
 		http.Error(w, msg, http.StatusInternalServerError)
 		LogError.Error(msg)
 		return
 	}
```

In this case, `gaurun` still returns HTTP OK(200) instead of Internal Server Error(500) even if `json.NewEncoder(w).Encode(&respGaurun)` fails. Because HTTP status code of response is fixed when `w.WriteHeader(code)` is called. ( multiple `WriteHeader()` call is not allowed. )

In addition, `WriteHeader()` is called in `json.NewEncoder(w).Encode(&respGaurun)` without up-front `w.WriteHeader(code)`.

To recap, it looks that the approach to giving `http.ResponseWriter` to `json.NewEncoder` directly cannot handle the error correctly.

So I'll merge the part of switch from `Unmarshal` to `NewDecoder` at present.

**Comment by mattn** [comment]
Considering cases of that `json.NewEncoder(w).Encode(&respGaurun)` fails. The respGaurun is simple structure to encode as JSON. so It may fail less. like below.

* Memory Allocation Error
  If it will occured, calling `json.Marshal` will be same. On the contrary `json.Marshal` allocate memory more than` json.NewEncoder (x) .Encode (& y)` since return encoded byte-s.

* Bug of `encoding/json`
 If it will be occured, calling `json.Marshal` will be same.

* Failed to write response (connection closed?)
  If it will occured, calling `fmt.Fprintf(w, string(respBody))` will be same.

**Comment by cubicdaiya** [comment]
@mattn I see. After minor changed, I'll merge your PR.

**Comment by cubicdaiya** [comment]
Minor changed and Merged. Thanks. https://github.com/mercari/gaurun/commit/2e5e2f6ad12b20cfe12c0ee029cbbd448892e533

**Comment by mattn** [comment]
Thank you
